File: day_02/day02.jl
#!/usr/env julia
# Advent of Code 2019
# Day 2

task_input = split.(readlines("input"),",")
data = collect(Iterators.flatten([parse.(Int, element) for element in task_input]))

operator = Dict(1 => +,
                2 => *)


function intcode!(noun, verb)
    num = deepcopy(data)
    num[2] = noun
    num[3] = verb

    let i = 1
        while num[i] ≠ 99 && i ≤ length(num)
            num[num[i+3]+1] = operator[num[i]](num[num[i+1]+1], num[num[i+2]+1])
            i += 4
        end
    end
    return num[1]
end

function intcode_iterate(stop_value)
    for i = 1:100*100
        noun, verb = divrem(i-1, 100) .+ (1,1)
        if intcode!(noun, verb) == stop_value
            return 100 * noun + verb
        end
    end
end

println(intcode!(12, 2)) # Task 1
println(intcode_iterate(19690720)) # Task 2

File: day02.py
#!/usr/bin/python3
# Advent of Code 2019
# Day 2

import operator

task_input = open("input/02").read().rstrip().split(",")
task_input = [int(element) for element in task_input]
operator = {1: operator.add,
            2: operator.mul}


def intcode(noun: int, verb: int):
    num = task_input.copy()
    num[1] = noun
    num[2] = verb
    i = 0
    while num[i] != 99 and i <= len(num):
        num[num[i+3]] = operator[num[i]](num[num[i+1]], num[num[i+2]])
        i += 4
    return num[0]


def intcode_iterate(stop_value: int):
    for noun in range(100):
        for verb in range(100):
            val = 100 * noun + verb if intcode(noun, verb) == stop_value else None
            if val is not None:
                return val


if __name__ == "__main__":
    # Task 1
    print(intcode(12, 2))

    # Task 2
    print(intcode_iterate(19690720))
